sstables/mx/partition_reversing_data_source.cc

```diff
@@ -0,0 +1,442 @@
+        }
+    }
+public:
+    row_body_skipping_context(input_stream<char>&& input, uint64_t start, uint64_t maxlen, reader_permit permit, column_translation ct)
+                : continuous_data_consumer(std::move(permit), std::move(input), start, maxlen)
+                , _gen(do_process_state())
+                , _column_translation(std::move(ct))
+    {}
+};
+
+class partition_reversing_data_source_impl final : public data_source_impl {
+    const schema& _schema;
+    shared_sstable _sst;
+    index_reader& _ir;
+    const ::io_priority_class& _io_priority;
+    reader_permit _permit;
+    tracing::trace_state_ptr _trace_state;
+    std::optional<partition_header_context> _partition_header_context;
+    std::optional<row_body_skipping_context> _row_reversing_context;
+    uint64_t _last_entry_pos;
+    uint64_t _clustering_range_start;
+    uint64_t _partition_start;
+    uint64_t _partition_end;
+    uint64_t _row_start;
+    uint64_t _row_end;
+
+    temporary_buffer<char> _cached_read;
+    uint64_t _current_read_size = 4 * 1024;
+    static const uint64_t max_read_size = 128 * 1024;
+
+    enum class state {
+        RANGE_END,
+        ROWS,
+        PARTITION_END,
+        PARTITION_ENDED
+    } _state = state::RANGE_END;
+private:
+    input_stream<char> data_stream(size_t start, size_t end) {
+        return _sst->data_stream(start, end - start, _io_priority, _permit, _trace_state, {});
+    }
+    future<temporary_buffer<char>> data_read(uint64_t start, uint64_t end) {
+        return _sst->data_read(start, end - start, _io_priority, _permit);
+    }
+    future<input_stream<char>> last_bytes_stream(size_t len) {
+        if (_cached_read.size() < len) {
+            if (_clustering_range_start + _current_read_size < _row_end) {
+                _cached_read = co_await data_read(std::min(_row_end - _current_read_size, _row_start), _row_end);
+            } else {
+                _cached_read = co_await data_read(_clustering_range_start, _row_end);
+            }
+            _current_read_size = std::min(max_read_size, _current_read_size * 2);
+        }
+        co_return make_buffer_input_stream(_cached_read.share(_cached_read.size() - len, len));
+    }
+    temporary_buffer<char> last_bytes(size_t len) {
+        auto tmp = _cached_read.share(_cached_read.size() - len, len);
+        _cached_read.trim(_cached_read.size() - len);
+        return tmp;
+    }
+
+    column_translation get_column_translation() {
+        return _sst->get_column_translation(_schema, _sst->get_serialization_header(), _sst->features());
+    }
+    void modify_cache(const row_body_skipping_context::tombstone_reversing_info& info) {
+        auto to_idx = [this] (uint64_t offset) {
+            return _cached_read.size() - (_row_end - offset);
+        };
+        _cached_read.get_write()[to_idx(info.kind_offset)] = 7 - (int)info.range_tombstone_kind;
+        if ((int)info.range_tombstone_kind < 2 || (int)info.range_tombstone_kind > 5) {
```

Comment:
A lot more comments are needed here.

Reply:
I think I'll handle this with a switch instead of converting the enums to ints